MovieTime/MovieTime/Service/MTService.swift

```diff
@@ -0,0 +1,205 @@
+            completion?(mtResponse)
+            return
+        }
+
+        guard let req = self.req(url: strURL, method: method, parameters: parameters, encoding: encoding, headers: headers) else {
+            let error = MTNetworkError.unexpectedError(message: "The service is unavailable.")
+            let mtResponse = MTResponse(statusCode: 500, data: nil, error: error)
+
+            MTLog(from: "REQ-ERROR", title: "\(error.message)")
+            completion?(mtResponse)
+            return
+        }
+
+        perform(request: req, completion: completion)
+    }
+
+    private func req(url: String, method: MTHTTPMethod = .get, parameters: MTParameters? = nil, encoding: MTParameterEncoding, headers: MTHTTPHeaders? = nil) -> URLRequest? {
+
+        guard let url = encoding.urlEncode(url, method: method, with: parameters) else { return nil }
+
+        // URL
+        var originalRequest = URLRequest(url: url)
+
+        // Method
+        originalRequest.httpMethod = method.rawValue
+
+        // Headers
+        for (key, value) in (headers ?? [:]) {
+            originalRequest.addValue(value, forHTTPHeaderField: key)
+        }
+
+        // Encoding
+        let encodedURLRequest = encoding.encode(originalRequest, with: parameters)
+
+        return encodedURLRequest
+    }
+
+    private func perform(request: URLRequest, completion: MTRequestResponse) {
+        MTLog(from: "REQUEST", title: "URL: \(request.url?.absoluteString ?? "NO URL!")")
+
+        let session = URLSession.shared
+        session.dataTask(with: request) { data, response, error in
+
+            let parsedError = self.parseError(data, response, error)
+
+            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 500
+            let mtResponse = MTResponse(statusCode: statusCode, data: data, error: parsedError, request: request, response: response)
+
+            if let data = mtResponse.data {
+                MTLog(from: "RESPONSE", title: "Data: \(String(data: data, encoding: .utf8) ?? "")")
+            }
+
+            if let error = mtResponse.error {
+                MTLog(from: "REQ-ERROR", title: "\(error.message)")
+            }
+            completion?(mtResponse)
+            }.resume()
+    }
+
+    private func parseError(_ data: Data?, _ response: URLResponse?, _ error: Error?) -> MTNetworkError? {
+        if let error = errorFrom(error) {
+            return error
+        }
+        if let responseError = errorFrom(response, data) {
+            return responseError
+        }
+
+        return nil
+    }
+
+    private func errorFrom(_ error: Error?) -> MTNetworkError? {
+        guard let nsError = error as NSError? else {
+            return nil
+        }
+
+        let systemErrors = [NSURLErrorNotConnectedToInternet,
+                            NSURLErrorUnknown]
+
+        if systemErrors.contains(nsError.code) {
+            return MTNetworkError.noConnection()
```

Comment:
is there a better mapping for `NSURLErrorUnknown` other than "no connection"?

Reply:
Yes, in this context the `NSURLErrorUnknown` could be a `MTNetworkError.unexpectedError(message: nsError.localizedDescription)`
Since the documentation has the following information:
```
Returned when the URL Loading system encounters an error that it cannot interpret.
```